### PR TITLE
Add auto balance command

### PR DESCRIPTION
**Pull request by Tyson453**

### CODE REVIEW

**Comment by NolanCampbell2357** [comment]
Make it a draft if it's in progress

**Comment by NolanCampbell2357** [comment]
@Tyson453 The rebase we did Yesterday was unsuccessful. We needed to do a `onto` rebase. I have fixed it. You will need to do a `git pull --rebase` to update your branch.

File: robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java [hunk 1]
```diff
@@ -264,6 +279,11 @@ public void drive(ChassisSpeeds chassisSpeeds) {
 		Logger.getInstance().recordOutput("Swerve States", states);
 	}
 
+	public boolean isBalanced() {
+		// return (-2.5 <= getRoll() && getRoll() <= 2.5) && (-2.5 <= getPitch() && getPitch() <= 2.5);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java:283]: Remove unused code

File: robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+        m_error = Math.copySign(Constants.DRIVE.BALANCE_LEVEL_DEGREES + Math.abs(m_angle), m_angle);
+        m_power = Math.min(Math.abs(Constants.DRIVE.BALANCE_KP * m_error), Constants.DRIVE.BALANCE_MAX_POWER);
+        m_power = Math.copySign(m_power, m_error);
+
+        m_power *= m_direction;
+
+        SwerveDriveSubsystem.getInstance().drive(m_power, 0, 0);
+
+        // System.out.println("Yaw: " + m_yaw);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java:49]: Remove unused code

File: robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Override
+    public void initialize() {
+        SwerveDriveSubsystem.getInstance().zero();
+    }
+
+    @Override
+    public void execute() {
+
+        m_yaw = Math.abs(SwerveDriveSubsystem.getInstance().getYaw() % 360);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java:25]: You can reduce wordiness by assigning a member variable like `m_swerve` to the swerve subsystem, and just refer to that variable.

Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java:25]: However, complex command logic should be moved to the subsystem. We want commands to have as little logic as possible, and do it in the subsystem instead.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]

File: robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java [hunk 4]
```diff
@@ -270,6 +284,39 @@ public void drive(ChassisSpeeds chassisSpeeds) {
...
+        } else if (225 <= yaw && yaw < 315) {
+            direction = -1;
+            angle = getPitch();
+        }
+
+		if (angle > Constants.DRIVE.BALANCE_FULL_TILT_DEGREES) {
+			return;
+		}
+
+		error = Math.copySign(Constants.DRIVE.BALANCE_LEVEL_DEGREES + Math.abs(angle), angle);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java:311]: I am on the fence about having these constants in subsystem. Usually, we want constants to be passed as arguments to functions, or to set them as configs on the subsystem. The swerve subsystem is a little unique in that we know we will reuse this code year to year, so I would be okay with having the constants used in this method only, just so it is easier to remove after the season. Then again, we intend to re-write this subsystem using our own swerve stuff anyway, so having configs might be better. Use your best judgement to decide.

File: robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        SwerveDriveSubsystem.getInstance().zero();
+    }
+
+    @Override
+    public void execute() {
+
+        SwerveDriveSubsystem.getInstance().balance();
+
+    }
+
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/commands/AutoBalanceCommand.java:28]: Even if the `isFinished` method always returns false, it is a good idea to override ride it, so it is visible.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]
Looks very good. Just a couple small things.

File: robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java (hunk 4 above)
Comment (Tyson453) [reply robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java:311]: Should I just move the balancing back into the command and put the if statements for the direction into a util class? Then there wouldn't be too much logic in the AutoBalanceCommand and there wouldn't be code specific to this year only in the SwerveDriveSubsystem.

Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/frc2023/subsystems/SwerveDriveSubsystem.java:311]: Let's just leave it for now

**Review by NolanCampbell2357 — APPROVED** [review]

**Pull request merged**